File: 2023/day1/Python/part2.py
import re


inp = open('../input.txt', 'r').readlines()

num_words = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
pattern = '\d|one|two|three|four|five|six|seven|eight|nine'
calib_sum = 0


for line in inp[:-1]:
    matched_nums = re.findall(pattern, line)
    first = matched_nums[0]
    last = matched_nums[-1]


    if (matched_nums[0] in num_words):
        first = str( num_words.index(matched_nums[0]) + 1 )

    if (matched_nums[-1] in num_words):
        possible_last_num = matched_nums[-1]

        while True:
            next_possible_match_idx = line.rfind(possible_last_num) + len(possible_last_num) - 1
            match = re.match(pattern, line[next_possible_match_idx:])

            if not match:
                break

            possible_last_num = match.group(0)

        last = str( num_words.index(possible_last_num) + 1 )


    calibration_value = first + last
    calib_sum += int(calibration_value)


print(calib_sum)

File: 2023/day01/Lua/part2.lua
local lines = io.open('../input.txt', 'r'):lines()

local num_conversions = {
    ['one']   = 'o1e',
    ['two']   = 't2o',
    ['three'] = 't3e',
    ['four']  = 'f4r',
    ['five']  = 'f5e',
    ['six']   = 's6x',
    ['seven'] = 's7n',
    ['eight'] = 'e8t',
    ['nine']  = 'n9e'
}

local numbers = {'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'}
local calib_sum = 0


for line in lines do
    if line ~= '' then
        -- search the line for each number and if there's a match,
        -- use the match as the key in the num_conversions table
        -- and replace the match with the value of that key
        for i = 1, #numbers do
            line = string.gsub(line, numbers[i], num_conversions)
        end


        local first = string.match(line,           '%d')
        local last  = string.match(line:reverse(), '%d')

        calib_sum = calib_sum + (first .. last)
    end
end


io.write(calib_sum)